keps/sig-apps/20190318-PodDisruptionBudget-graduation-to-stable.md

```diff
@@ -0,0 +1,132 @@
+---
+title: graduate-pod-disruption-budget-to-stable
+authors:
+  - "@bsalamat"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-scheduling
+reviewers:
+  - "@liggitt"
+  - "@kow3ns"
+approvers:
+  - "@kow3ns"
+  - "@liggitt"
+editor: TBD
+creation-date: 2019-03-18
+last-updated: 2019-03-18
+status: implementable
+see-also:
+  - 
+replaces:
+superseded-by:
+---
+
+# Graduate PodDisruptionBudget to stable
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+[Pod Disruption Budget (PDB)](https://kubernetes.io/docs/tasks/run-application/configure-pdb/)
+is a Kubernetes API that limits the number of pods of a collection that are down simultaneously from voluntary disruptions.
+[Kubernetes eviction API](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#the-eviction-api)
+takes PDB into account when terminating pods. If PDB is
+violated, the eviction API returns failure and does not delete the requested pod.
+The feature was introduced in Kubernetes 1.4 and promoted to beta in 1.5.
+It has been in beta for a long time. This document lays out the plan to promote
+it to stable.
+
+## Motivation
+
+PDB API has been stable and is an important feature that allows users to improve
+reliability of their critical workloads. This feature has been in beta for a
+long time. We need to promote to stable version given that we plan to support it
+long term.
+
+### Goals
+
+* Plan to promote PDB API to stable version.
+* List action items to mitigate potential risks of supporting mutable PDB.
+
+### Non-Goals
+
+* Making changes to the API fields or their meanings.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### Mutable PDBs
+
+A mutable PDB object allows its `MinAvailable`, `MaxUnavailable`, and `Selector`
+fields to be modified by clients. Components that use PDB must watch such
+modifications and use the updated values when making decisions.
+
+This feature is implemented by [this PR](https://github.com/kubernetes/kubernetes/pull/69867).
+
+### Risks and Mitigations
+
+We plan to support mutation of PDB objects that didn't exist in previous versions.
+The following needs to be checked and verified before graduation of the API.
+
+- [ ] Ensure that components do not have any logic that relies on immutability
+of PDBs. For example, if component builds a cache of pods that match various
+PDBs, they must add logic to invalid the cache on updates.
+   - Action Item: sweep for in-tree uses to make sure we’re informer driven and
```

Comment:
This should include all repos in kubernetes org, not just k/k. I know Cluster Autoscaler has logic to deal with PDBs (I don't think it will be impacted by this change, but it's worth double checking), I suspect VPA may also use it. Not sure about other projects under kubernetes org.

Reply:
Added